### PR TITLE
Fixes #4335: Addition of support for uploading performance metric logs 

### LINKED ISSUE #4335: Add log upload support for performance metrics collection
This issue tracks the implementation of support for uploading log reports related to performance metrics collection to the remote service.

### PR DESCRIPTION
**Pull request by Sarthak2601**
## Explanation

Fixes #4335
- Introduces support for uploading performance metric logs to the remote service. 
- The PR modifies files related to work manager to add and queue work request for uploading metric logs. It also modifies files related to unpacking and repacking of event logs into bundles for logging to firebase. 
- Added tests to verify these functionalities as well. 

## Essential Checklist

- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

### CODE REVIEW

File: domain/src/main/java/org/oppia/android/domain/oppialogger/loguploader/LogUploadWorkManagerInitializer.kt
Comment (BenHenning) [inline domain/src/main/java/org/oppia/android/domain/oppialogger/loguploader/LogUploadWorkManagerInitializer.kt]: Is there a test to verify this is working correctly?

File: domain/src/main/java/org/oppia/android/domain/oppialogger/loguploader/LogUploadWorker.kt
Comment (BenHenning) [inline domain/src/main/java/org/oppia/android/domain/oppialogger/loguploader/LogUploadWorker.kt]: What purpose does using 'let' serve here? It might be decreasing the readability of the code since it seems unnecessary.

File: utility/src/main/java/org/oppia/android/util/logging/firebase/LogReportingModule.kt
Comment (BenHenning) [inline utility/src/main/java/org/oppia/android/util/logging/firebase/LogReportingModule.kt]: Any test to verify this binding?

File: domain/src/main/java/org/oppia/android/domain/oppialogger/loguploader/LogUploadWorker.kt
Comment (BenHenning) [inline domain/src/main/java/org/oppia/android/domain/oppialogger/loguploader/LogUploadWorker.kt:55]: Any tests to verify that this is working as expected?

File: utility/src/main/java/org/oppia/android/util/logging/EventBundleCreator.kt
Comment (BenHenning) [inline utility/src/main/java/org/oppia/android/util/logging/EventBundleCreator.kt]: A nit on member ordering: since the method above appears after the other public API method in this class, I feel like this helper should appear after its helper (since it appears next in the class order).

File: utility/src/main/java/org/oppia/android/util/logging/EventBundleCreator.kt
Comment (BenHenning) [inline utility/src/main/java/org/oppia/android/util/logging/EventBundleCreator.kt]: Can probably just be: ``No performance metrics need to be tied to user IDs.`` since I don't think we have long-term plans for this, either.

File: utility/src/main/java/org/oppia/android/util/logging/EventBundleCreator.kt
Comment (BenHenning) [inline utility/src/main/java/org/oppia/android/util/logging/EventBundleCreator.kt:113]: Please add a KDoc for this method since it's an addition to the class's public API.

Comment (BenHenning) [inline utility/src/main/java/org/oppia/android/util/logging/EventBundleCreator.kt:113]: Any tests for this new functionality?

File: utility/src/main/java/org/oppia/android/util/logging/firebase/FirebaseEventLogger.kt
Comment (BenHenning) [inline utility/src/main/java/org/oppia/android/util/logging/firebase/FirebaseEventLogger.kt]: How painful would it be to rename the existing event logger? I'm just a bit concerned that it's too general sounding now that we're specializing performance metrics.

File: utility/src/main/java/org/oppia/android/util/logging/firebase/FirebaseEventLogger.kt
Comment (BenHenning) [inline utility/src/main/java/org/oppia/android/util/logging/firebase/FirebaseEventLogger.kt]: Please add a KDoc for this. Also, any new tests for this method?

**Review by BenHenning — COMMENTED** [review]
Thanks @Sarthak2601. Took a first pass--PTAL.

File: utility/src/main/java/org/oppia/android/util/logging/firebase/FirebaseEventLogger.kt
Comment (Sarthak2601) [reply utility/src/main/java/org/oppia/android/util/logging/firebase/FirebaseEventLogger.kt]: We'll need to do the following: 
- changes in all the production files where eventLogger is used,
- changes in test files. 
- changes in dagger modules. 
- refactoring of fakeEventlogger.

It might not be too painful -- we can file this issue as a good-second-issue.

File: utility/src/main/java/org/oppia/android/util/logging/firebase/FirebaseEventLogger.kt
Comment (Sarthak2601) [reply utility/src/main/java/org/oppia/android/util/logging/firebase/FirebaseEventLogger.kt]: added kdoc.

File: utility/src/main/java/org/oppia/android/util/logging/EventBundleCreator.kt
Comment (Sarthak2601) [reply utility/src/main/java/org/oppia/android/util/logging/EventBundleCreator.kt:113]: done.

File: utility/src/main/java/org/oppia/android/util/logging/EventBundleCreator.kt
Comment (Sarthak2601) [reply utility/src/main/java/org/oppia/android/util/logging/EventBundleCreator.kt]: done.

File: utility/src/main/java/org/oppia/android/util/logging/EventBundleCreator.kt
Comment (Sarthak2601) [reply utility/src/main/java/org/oppia/android/util/logging/EventBundleCreator.kt]: done.

File: domain/src/main/java/org/oppia/android/domain/oppialogger/loguploader/LogUploadWorker.kt
Comment (Sarthak2601) [reply domain/src/main/java/org/oppia/android/domain/oppialogger/loguploader/LogUploadWorker.kt]: replaced with forEach to enhance readability.

File: domain/src/main/java/org/oppia/android/domain/oppialogger/loguploader/LogUploadWorkManagerInitializer.kt
Comment (Sarthak2601) [reply domain/src/main/java/org/oppia/android/domain/oppialogger/loguploader/LogUploadWorkManagerInitializer.kt]: added.

File: utility/src/main/java/org/oppia/android/util/logging/firebase/LogReportingModule.kt
Comment (Sarthak2601) [reply utility/src/main/java/org/oppia/android/util/logging/firebase/LogReportingModule.kt]: I've added tests to verify this binding with the help of a fake in the earlier PR. Does that suffice?

File: domain/src/main/java/org/oppia/android/domain/oppialogger/loguploader/LogUploadWorker.kt
Comment (Sarthak2601) [reply domain/src/main/java/org/oppia/android/domain/oppialogger/loguploader/LogUploadWorker.kt:55]: added.

File: utility/src/main/java/org/oppia/android/util/logging/EventBundleCreator.kt
Comment (Sarthak2601) [reply utility/src/main/java/org/oppia/android/util/logging/EventBundleCreator.kt:113]: added.

File: utility/src/main/java/org/oppia/android/util/logging/firebase/FirebaseEventLogger.kt
Comment (Sarthak2601) [reply utility/src/main/java/org/oppia/android/util/logging/firebase/FirebaseEventLogger.kt]: Filed an issue for this -- https://github.com/oppia/oppia-android/issues/4405

File: utility/src/test/java/org/oppia/android/util/logging/EventBundleCreatorTest.kt
Comment (BenHenning) [inline utility/src/test/java/org/oppia/android/util/logging/EventBundleCreatorTest.kt]: ``eventWithDefaultloggableMetric``? There's no "context" field name in this proto.

File: utility/src/main/java/org/oppia/android/util/logging/firebase/LogReportingModule.kt
Comment (BenHenning) [reply utility/src/main/java/org/oppia/android/util/logging/firebase/LogReportingModule.kt]: I suggest adding an explicit LogReportingModuleTest (see other module tests for an idea)--they're generally quick to setup, and are useful for directly verifying the exact binding.

File: utility/src/main/java/org/oppia/android/util/logging/firebase/FirebaseEventLogger.kt
Comment (BenHenning) [reply utility/src/main/java/org/oppia/android/util/logging/firebase/FirebaseEventLogger.kt]: That seems reasonable--thanks for filing the issue!

**Review by BenHenning — COMMENTED** [review]
Thanks @Sarthak2601! Just had a few comments--PTAL.

File: utility/src/test/java/org/oppia/android/util/logging/EventBundleCreatorTest.kt
Comment (Sarthak2601) [reply utility/src/test/java/org/oppia/android/util/logging/EventBundleCreatorTest.kt]: done.

File: utility/src/main/java/org/oppia/android/util/logging/firebase/LogReportingModule.kt
Comment (Sarthak2601) [reply utility/src/main/java/org/oppia/android/util/logging/firebase/LogReportingModule.kt]: done.

**Comment by Sarthak2601** [comment]
@BenHenning if this PR looks good to you, please don't merge after approval -- we'll merge the stacked PRs one by one otherwise all file changes will get stuffed up in one PR which will then become very tedious to review and manage :)

File: utility/src/test/java/org/oppia/android/util/logging/EventBundleCreatorTest.kt
Comment (BenHenning) [inline utility/src/test/java/org/oppia/android/util/logging/EventBundleCreatorTest.kt]:
```suggestion
  fun testFillMetricsBundle_eventWithDefaultLoggableMetric_fillsDetailsAndRetsUnknownLog() {
```

Maybe?

File: utility/src/test/java/org/oppia/android/util/logging/firebase/LogReportingModuleTest.kt
Comment (BenHenning) [inline utility/src/test/java/org/oppia/android/util/logging/firebase/LogReportingModuleTest.kt]: Nit: please directly import assertThat for parity with other test suites.

File: utility/src/test/java/org/oppia/android/util/logging/firebase/LogReportingModuleTest.kt
Comment (BenHenning) [inline utility/src/test/java/org/oppia/android/util/logging/firebase/LogReportingModuleTest.kt:60]: Thanks for adding a test for the existing binding as well!

**Review by BenHenning — COMMENTED** [review]
Thanks @Sarthak2601. Just 2 nits, otherwise LGTM.

**Comment by BenHenning** [comment]
> @BenHenning if this PR looks good to you, please don't merge after approval -- we'll merge the stacked PRs one by one otherwise all file changes will get stuffed up in one PR which will then become very tedious to review and manage :)

I agree--I generally only merge branches that go into ``develop``.

Also, it looks like a lot of CI checks are failing, so I slightly withdraw my previous "LGTM" until these are also addressed so that the PR is green.

File: utility/src/test/java/org/oppia/android/util/logging/EventBundleCreatorTest.kt
Comment (Sarthak2601) [reply utility/src/test/java/org/oppia/android/util/logging/EventBundleCreatorTest.kt]: done.

File: utility/src/test/java/org/oppia/android/util/logging/firebase/LogReportingModuleTest.kt
Comment (Sarthak2601) [reply utility/src/test/java/org/oppia/android/util/logging/firebase/LogReportingModuleTest.kt]: done.

**Comment by Sarthak2601** [comment]
> > @BenHenning if this PR looks good to you, please don't merge after approval -- we'll merge the stacked PRs one by one otherwise all file changes will get stuffed up in one PR which will then become very tedious to review and manage :)
> 
> I agree--I generally only merge branches that go into `develop`.
> 
> Also, it looks like a lot of CI checks are failing, so I slightly withdraw my previous "LGTM" until these are also addressed so that the PR is green.

@BenHenning I've fixed all CI tests, PTAL -- thanks!

**Review by BenHenning — COMMENTED** [review]
Thanks @Sarthak2601! The PR generally LGTM. Will wait to approve after the base PR is merged and this one is updated to be based on develop.

**Review by BenHenning — APPROVED** [review]
Thanks @Sarthak2601! LGTM. Feel free to merge whenever you need to.

**Comment by BenHenning** [comment]
(Unassigning Rajat since this PR doesn't need his review).

**Pull request merged**
